File: bench/scala/src/main/scala/ming/Parser.scala
package ming

private[ming] case class Pos(line: Int, col: Int):
  override def toString: String = s"$line:$col"

private[ming] enum Expr:
  case Num(value: Long, pos: Pos = Pos(0, 0))
  case Bool(value: Boolean, pos: Pos = Pos(0, 0))
  case Str(value: String, pos: Pos = Pos(0, 0))
  case Symbol(name: String, pos: Pos = Pos(0, 0))
  case SList(elems: List[Expr], pos: Pos = Pos(0, 0))

private[ming] object Parser:

  private case class Token(text: String, pos: Pos)

  private def tokenize(input: String): List[Token] =
    val tokens = scala.collection.mutable.ListBuffer[Token]()
    var i      = 0
    var line   = 1
    var col    = 1
    while i < input.length do
      input(i) match
        case '\n' =>
          i += 1; line += 1; col = 1
        case c if c.isWhitespace =>
          i += 1; col += 1
        case ';' =>
          val next = skipComment(input, i)
          col += (next - i)
          i = next
        case '(' | ')' | '\'' =>
          tokens += Token(input(i).toString, Pos(line, col))
          i += 1; col += 1
        case '"' =>
          val p           = Pos(line, col)
          val (tok, next) = readString(input, i)
          tokens += Token(tok, p)
          col += (next - i)
          i = next
        case '#' =>
          val p           = Pos(line, col)
          val (tok, next) = readHash(input, i)
          tokens += Token(tok, p)
          col += (next - i)
          i = next
        case _ =>
          val p           = Pos(line, col)
          val (tok, next) = readWord(input, i)
          tokens += Token(tok, p)
          col += (next - i)
          i = next
    tokens.toList

  def parseAll(input: String): List[Expr] =
    val tokens = tokenize(input)
    val exprs  = scala.collection.mutable.ListBuffer[Expr]()
    var rest   = tokens
    while rest.nonEmpty do
      val (expr, remaining) = parseExpr(rest)
      exprs += expr
      rest = remaining
    exprs.toList

  private def parseExpr(tokens: List[Token]): (Expr, List[Token]) = tokens match
    case Nil => throw EvalError("unexpected end of input")
    case Token("(", p) :: rest =>
      val (elems, remaining) = parseList(rest)
      (Expr.SList(elems, p), remaining)
    case Token("'", p) :: rest =>
      val (expr, remaining) = parseExpr(rest)
      (Expr.SList(List(Expr.Symbol("quote", p), expr), p), remaining)
    case Token(")", _) :: _     => throw EvalError("unexpected )")
    case Token(text, p) :: rest => (parseAtom(text, p), rest)

  private def parseList(tokens: List[Token]): (List[Expr], List[Token]) =
    val elems = scala.collection.mutable.ListBuffer[Expr]()
    var rest  = tokens
    while rest.nonEmpty && rest.head.text != ")" do
      val (expr, remaining) = parseExpr(rest)
      elems += expr
      rest = remaining
    if rest.isEmpty then throw EvalError("missing )")
    (elems.toList, rest.tail)

  private def parseAtom(token: String, pos: Pos): Expr =
    if token == "#t" then Expr.Bool(true, pos)
    else if token == "#f" then Expr.Bool(false, pos)
    else if token.startsWith("\"") then Expr.Str(token.substring(1, token.length - 1), pos)
    else
      token.toLongOption match
        case Some(n) => Expr.Num(n, pos)
        case None    => Expr.Symbol(token, pos)

  private def readWord(input: String, start: Int): (String, Int) =
    val sb = new StringBuilder
    var i  = start
    while i < input.length && !input(i).isWhitespace && input(i) != '(' && input(i) != ')' do
      sb += input(i); i += 1
    (sb.toString, i)

  private def readString(input: String, start: Int): (String, Int) =
    val sb = new StringBuilder("\"")
    var i  = start + 1
    while i < input.length && input(i) != '"' do
      if input(i) == '\\' then
        sb += input(i); i += 1
        if i < input.length then
          sb += input(i); i += 1
      else
        sb += input(i); i += 1
    if i < input.length then
      sb += '"'; i += 1
    (sb.toString, i)

  private def readHash(input: String, start: Int): (String, Int) =
    if start + 1 >= input.length then ("#", start + 1)
    else
      val next = input(start + 1)
      if next == 't' || next == 'f' then (input.substring(start, start + 2), start + 2)
      else readWord(input, start)

  private def skipComment(input: String, start: Int): Int =
    var i = start
    while i < input.length && input(i) != '\n' do i += 1
    i

File: bench/java/src/main/java/ming/Parser.java
package ming;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.List;

final class Parser {
    private final String input;
    private int index;
    private int line;
    private int column;

    Parser(String input) {
        this.input = input;
        this.index = 0;
        this.line = 1;
        this.column = 1;
    }

    List<Expr> parseProgram() throws EvalError {
        List<Expr> expressions = new ArrayList<>();
        skipIgnored();
        while (!isAtEnd()) {
            expressions.add(parseExpression());
            skipIgnored();
        }
        return List.copyOf(expressions);
    }

    private Expr parseExpression() throws EvalError {
        skipIgnored();
        if (isAtEnd()) {
            throw new EvalError("unexpected end of input");
        }

        SourceLoc loc = currentLoc();
        char ch = peek();
        if (ch == '\'') {
            return parseQuoteShorthand();
        }
        if (ch == '(') {
            return parseList();
        }
        if (ch == ')') {
            throw SchemeErrors.at(loc, "unexpected ')'");
        }
        if (ch == '"') {
            return parseString();
        }
        return parseAtom();
    }

    private Expr parseList() throws EvalError {
        SourceLoc start = currentLoc();
        advance();

        List<Expr> elements = new ArrayList<>();
        skipIgnored();
        while (!isAtEnd() && peek() != ')') {
            elements.add(parseExpression());
            skipIgnored();
        }

        if (isAtEnd()) {
            throw SchemeErrors.at(start, "unterminated list");
        }
        advance();
        return new ListExpr(start, List.copyOf(elements));
    }

    private Expr parseQuoteShorthand() throws EvalError {
        SourceLoc start = currentLoc();
        advance();
        Expr quotedExpr = parseExpression();
        return new ListExpr(start, List.of(new SymbolExpr(start, "quote"), quotedExpr));
    }

    private Expr parseString() throws EvalError {
        SourceLoc start = currentLoc();
        advance();

        StringBuilder builder = new StringBuilder();
        while (!isAtEnd()) {
            char ch = advance();
            if (ch == '"') {
                return new StringExpr(start, builder.toString());
            }
            if (ch == '\\') {
                if (isAtEnd()) {
                    throw SchemeErrors.at(start, "unterminated string literal");
                }
                builder.append(parseEscape(advance()));
            } else {
                builder.append(ch);
            }
        }

        throw SchemeErrors.at(start, "unterminated string literal");
    }

    private char parseEscape(char escaped) {
        return switch (escaped) {
            case 'n' -> '\n';
            case 'r' -> '\r';
            case 't' -> '\t';
            case '"' -> '"';
            case '\\' -> '\\';
            default -> escaped;
        };
    }

    private Expr parseAtom() throws EvalError {
        SourceLoc start = currentLoc();
        StringBuilder builder = new StringBuilder();
        while (!isAtEnd() && !isDelimiter(peek())) {
            builder.append(advance());
        }

        String token = builder.toString();
        if ("#t".equals(token)) {
            return new BooleanExpr(start, true);
        }
        if ("#f".equals(token)) {
            return new BooleanExpr(start, false);
        }
        if (token.startsWith("#\\")) {
            return new CharExpr(start, parseCharacterLiteral(token, start));
        }
        if (Rational.isIntegerToken(token)) {
            return new NumberExpr(start, Rational.integer(new BigInteger(token)));
        }
        return new SymbolExpr(start, token);
    }

    private void skipIgnored() {
        while (!isAtEnd()) {
            char ch = peek();
            if (Character.isWhitespace(ch)) {
                advance();
                continue;
            }
            if (ch == ';') {
                skipComment();
                continue;
            }
            return;
        }
    }

    private void skipComment() {
        while (!isAtEnd() && peek() != '\n') {
            advance();
        }
    }

    private boolean isDelimiter(char ch) {
        return Character.isWhitespace(ch) || ch == '(' || ch == ')' || ch == ';';
    }

    private boolean isAtEnd() {
        return index >= input.length();
    }

    private char peek() {
        return input.charAt(index);
    }

    private char advance() {
        char ch = input.charAt(index);
        index++;
        if (ch == '\n') {
            line++;
            column = 1;
        } else {
            column++;
        }
        return ch;
    }

    private SourceLoc currentLoc() {
        return new SourceLoc(line, column);
    }

    private int parseCharacterLiteral(String token, SourceLoc loc) throws EvalError {
        String literal = token.substring(2);
        return switch (literal) {
            case "space" -> ' ';
            case "newline" -> '\n';
            default -> {
                if (literal.codePointCount(0, literal.length()) != 1) {
                    throw SchemeErrors.at(loc, "invalid character literal");
                }
                yield literal.codePointAt(0);
            }
        };
    }
}
